// discord-player/dist/Player.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Player = void 0;
const discord_js_1 = require("discord.js");
const tiny_typed_emitter_1 = require("tiny-typed-emitter");
const Queue_1 = require("./Structures/Queue");
const VoiceUtils_1 = require("./VoiceInterface/VoiceUtils");
const types_1 = require("./types/types");
const Track_1 = __importDefault(require("./Structures/Track"));
const QueryResolver_1 = require("./utils/QueryResolver");
const youtube_sr_1 = __importDefault(require("youtube-sr"));
const Util_1 = require("./utils/Util");
const spotify_url_info_1 = __importDefault(require("spotify-url-info"));
const PlayerError_1 = require("./Structures/PlayerError");
const ytdl_core_1 = require("ytdl-core");
const soundcloud_scraper_1 = require("soundcloud-scraper");
const Playlist_1 = require("./Structures/Playlist");
const ExtractorModel_1 = require("./Structures/ExtractorModel");
const voice_1 = require("@discordjs/voice");
const soundcloud = new soundcloud_scraper_1.Client();
class Player extends tiny_typed_emitter_1.TypedEmitter {
    /**
     * Creates new Discord Player
     * @param {Client} client The Discord Client
     * @param {PlayerInitOptions} [options={}] The player init options
     */
    constructor(client, options = {}) {
        var _a;
        super();
        this.options = {
            autoRegisterExtractor: true,
            ytdlOptions: {
                highWaterMark: 1 << 25
            },
            connectionTimeout: 20000
        };
        this.queues = new discord_js_1.Collection();
        this.voiceUtils = new VoiceUtils_1.VoiceUtils();
        this.extractors = new discord_js_1.Collection();
        /**
         * The discord.js client
         * @type {Client}
         */
        this.client = client;
        if (!new discord_js_1.Intents(this.client.options.intents).has(discord_js_1.Intents.FLAGS.GUILD_VOICE_STATES)) {
            throw new PlayerError_1.PlayerError('client is missing "GUILD_VOICE_STATES" intent');
        }
        /**
         * The extractors collection
         * @type {ExtractorModel}
         */
        this.options = Object.assign(this.options, options);
        this.client.on("voiceStateUpdate", this._handleVoiceState.bind(this));
        if ((_a = this.options) === null || _a === void 0 ? void 0 : _a.autoRegisterExtractor) {
            let nv; // eslint-disable-line @typescript-eslint/no-explicit-any
            if ((nv = Util_1.Util.require("@discord-player/extractor"))) {
                ["Attachment", "Facebook", "Reverbnation", "Vimeo"].forEach((ext) => void this.use(ext, nv[ext]));
            }
        }
    }
    /**
     * Handles voice state update
     * @param {VoiceState} oldState The old voice state
     * @param {VoiceState} newState The new voice state
     * @returns {void}
     * @private
     */
    _handleVoiceState(oldState, newState) {
        const queue = this.getQueue(oldState.guild.id);
        if (!queue)
            return;
        if (oldState.channelId && newState.channelId && oldState.channelId !== newState.channelId) {
            if (queue === null || queue === void 0 ? void 0 : queue.connection)
                queue.connection.channel = newState.channel;
        }
        if (!oldState.channelId && newState.channelId && newState.member.id === newState.guild.me.id) {
            if (newState.serverMute || !newState.serverMute) {
                queue.setPaused(newState.serverMute);
            }
            else if (newState.suppress || !newState.suppress) {
                if (newState.suppress)
                    newState.guild.me.voice.setRequestToSpeak(true).catch(Util_1.Util.noop);
                queue.setPaused(newState.suppress);
            }
        }
        if (oldState.channelId === newState.channelId && oldState.member.id === newState.guild.me.id) {
            if (oldState.serverMute !== newState.serverMute) {
                queue.setPaused(newState.serverMute);
            }
            else if (oldState.suppress !== newState.suppress) {
                if (newState.suppress)
                    newState.guild.me.voice.setRequestToSpeak(true).catch(Util_1.Util.noop);
                queue.setPaused(newState.suppress);
            }
        }
        if (oldState.member.id === this.client.user.id && !newState.channelId) {
            queue.destroy();
            return void this.emit("botDisconnect", queue);
        }
        if (!queue.connection || !queue.connection.channel)
            return;
        if (!oldState.channelId || newState.channelId) {
            const emptyTimeout = queue._cooldownsTimeout.get(`empty_${oldState.guild.id}`);
            const channelEmpty = Util_1.Util.isVoiceEmpty(queue.connection.channel);
            if (!channelEmpty && emptyTimeout) {
                clearTimeout(emptyTimeout);
                queue._cooldownsTimeout.delete(`empty_${oldState.guild.id}`);
            }
        }
        else {
            if (!Util_1.Util.isVoiceEmpty(queue.connection.channel))
                return;
            const timeout = setTimeout(() => {
                if (!Util_1.Util.isVoiceEmpty(queue.connection.channel))
                    return;
                if (!this.queues.has(queue.guild.id))
                    return;
                if (queue.options.leaveOnEmpty)
                    queue.destroy();
                this.emit("channelEmpty", queue);
            }, queue.options.leaveOnEmptyCooldown || 0).unref();
            queue._cooldownsTimeout.set(`empty_${oldState.guild.id}`, timeout);
        }
    }
    /**
     * Creates a queue for a guild if not available, else returns existing queue
     * @param {GuildResolvable} guild The guild
     * @param {PlayerOptions} queueInitOptions Queue init options
     * @returns {Queue}
     */
    createQueue(guild, queueInitOptions = {}) {
        var _a;
        guild = this.client.guilds.resolve(guild);
        if (!guild)
            throw new PlayerError_1.PlayerError("Unknown Guild", PlayerError_1.ErrorStatusCode.UNKNOWN_GUILD);
        if (this.queues.has(guild.id))
            return this.queues.get(guild.id);
        const _meta = queueInitOptions.metadata;
        delete queueInitOptions["metadata"];
        (_a = queueInitOptions.ytdlOptions) !== null && _a !== void 0 ? _a : (queueInitOptions.ytdlOptions = this.options.ytdlOptions);
        const queue = new Queue_1.Queue(this, guild, queueInitOptions);
        queue.metadata = _meta;
        this.queues.set(guild.id, queue);
        return queue;
    }
    /**
     * Returns the queue if available
     * @param {GuildResolvable} guild The guild id
     * @returns {Queue}
     */
    getQueue(guild) {
        guild = this.client.guilds.resolve(guild);
        if (!guild)
            throw new PlayerError_1.PlayerError("Unknown Guild", PlayerError_1.ErrorStatusCode.UNKNOWN_GUILD);
        return this.queues.get(guild.id);
    }
    /**
     * Deletes a queue and returns deleted queue object
     * @param {GuildResolvable} guild The guild id to remove
     * @returns {Queue}
     */
    deleteQueue(guild) {
        guild = this.client.guilds.resolve(guild);
        if (!guild)
            throw new PlayerError_1.PlayerError("Unknown Guild", PlayerError_1.ErrorStatusCode.UNKNOWN_GUILD);
        const prev = this.getQueue(guild);
        try {
            prev.destroy();
        }
        catch (_a) { } // eslint-disable-line no-empty
        this.queues.delete(guild.id);
        return prev;
    }
    /**
     * @typedef {object} SearchResult
     * @property {Playlist} [playlist] The playlist (if any)
     * @property {Track[]} tracks The tracks
     */
    /**
     * Search tracks
     * @param {string|Track} query The search query
     * @param {SearchOptions} options The search options
     * @returns {Promise<SearchResult>}
     */
    search(query, options) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18;
        return __awaiter(this, void 0, void 0, function* () {
            if (query instanceof Track_1.default)
                return { playlist: null, tracks: [query] };
            if (!options)
                throw new PlayerError_1.PlayerError("DiscordPlayer#search needs search options!", PlayerError_1.ErrorStatusCode.INVALID_ARG_TYPE);
            options.requestedBy = this.client.users.resolve(options.requestedBy);
            if (!("searchEngine" in options))
                options.searchEngine = types_1.QueryType.AUTO;
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            for (const [_, extractor] of this.extractors) {
                if (options.blockExtractor)
                    break;
                if (!extractor.validate(query))
                    continue;
                const data = yield extractor.handle(query);
                if (data && data.data.length) {
                    const playlist = !data.playlist
                        ? null
                        : new Playlist_1.Playlist(this, Object.assign(Object.assign({}, data.playlist), { tracks: [] }));
                    const tracks = data.data.map((m) => new Track_1.default(this, Object.assign(Object.assign({}, m), { requestedBy: options.requestedBy, duration: Util_1.Util.buildTimeCode(Util_1.Util.parseMS(m.duration)), playlist: playlist })));
                    if (playlist)
                        playlist.tracks = tracks;
                    return { playlist: playlist, tracks: tracks };
                }
            }
            const qt = options.searchEngine === types_1.QueryType.AUTO ? QueryResolver_1.QueryResolver.resolve(query) : options.searchEngine;
            switch (qt) {
                case types_1.QueryType.YOUTUBE_VIDEO: {
                    const info = yield ytdl_core_1.getInfo(query).catch(Util_1.Util.noop);
                    if (!info)
                        return { playlist: null, tracks: [] };
                    const track = new Track_1.default(this, {
                        title: info.videoDetails.title,
                        description: info.videoDetails.description,
                        author: (_a = info.videoDetails.author) === null || _a === void 0 ? void 0 : _a.name,
                        url: info.videoDetails.video_url,
                        requestedBy: options.requestedBy,
                        thumbnail: (_b = Util_1.Util.last(info.videoDetails.thumbnails)) === null || _b === void 0 ? void 0 : _b.url,
                        views: parseInt(info.videoDetails.viewCount.replace(/[^0-9]/g, "")) || 0,
                        duration: Util_1.Util.buildTimeCode(Util_1.Util.parseMS(parseInt(info.videoDetails.lengthSeconds) * 1000)),
                        source: "youtube",
                        raw: info
                    });
                    return { playlist: null, tracks: [track] };
                }
                case types_1.QueryType.YOUTUBE_SEARCH: {
                    const videos = yield youtube_sr_1.default.search(query, {
                        type: "video"
                    }).catch(Util_1.Util.noop);
                    if (!videos)
                        return { playlist: null, tracks: [] };
                    const tracks = videos.map((m) => {
                        var _a, _b;
                        m.source = "youtube"; // eslint-disable-line @typescript-eslint/no-explicit-any
                        return new Track_1.default(this, {
                            title: m.title,
                            description: m.description,
                            author: (_a = m.channel) === null || _a === void 0 ? void 0 : _a.name,
                            url: m.url,
                            requestedBy: options.requestedBy,
                            thumbnail: (_b = m.thumbnail) === null || _b === void 0 ? void 0 : _b.displayThumbnailURL("maxresdefault"),
                            views: m.views,
                            duration: m.durationFormatted,
                            source: "youtube",
                            raw: m
                        });
                    });
                    return { playlist: null, tracks };
                }
                case types_1.QueryType.SOUNDCLOUD_TRACK:
                case types_1.QueryType.SOUNDCLOUD_SEARCH: {
                    const result = QueryResolver_1.QueryResolver.resolve(query) === types_1.QueryType.SOUNDCLOUD_TRACK ? [{ url: query }] : yield soundcloud.search(query, "track").catch(() => []);
                    if (!result || !result.length)
                        return { playlist: null, tracks: [] };
                    const res = [];
                    for (const r of result) {
                        const trackInfo = yield soundcloud.getSongInfo(r.url).catch(Util_1.Util.noop);
                        if (!trackInfo)
                            continue;
                        const track = new Track_1.default(this, {
                            title: trackInfo.title,
                            url: trackInfo.url,
                            duration: Util_1.Util.buildTimeCode(Util_1.Util.parseMS(trackInfo.duration)),
                            description: trackInfo.description,
                            thumbnail: trackInfo.thumbnail,
                            views: trackInfo.playCount,
                            author: trackInfo.author.name,
                            requestedBy: options.requestedBy,
                            source: "soundcloud",
                            engine: trackInfo
                        });
                        res.push(track);
                    }
                    return { playlist: null, tracks: res };
                }
                case types_1.QueryType.SPOTIFY_SONG: {
                    const spotifyData = yield spotify_url_info_1.default.getData(query).catch(Util_1.Util.noop);
                    if (!spotifyData)
                        return { playlist: null, tracks: [] };
                    const spotifyTrack = new Track_1.default(this, {
                        title: spotifyData.name,
                        description: (_c = spotifyData.description) !== null && _c !== void 0 ? _c : "",
                        author: (_e = (_d = spotifyData.artists[0]) === null || _d === void 0 ? void 0 : _d.name) !== null && _e !== void 0 ? _e : "Unknown Artist",
                        url: (_g = (_f = spotifyData.external_urls) === null || _f === void 0 ? void 0 : _f.spotify) !== null && _g !== void 0 ? _g : query,
                        thumbnail: ((_k = (_j = (_h = spotifyData.album) === null || _h === void 0 ? void 0 : _h.images[0]) === null || _j === void 0 ? void 0 : _j.url) !== null && _k !== void 0 ? _k : (_l = spotifyData.preview_url) === null || _l === void 0 ? void 0 : _l.length)
                            ? `https://i.scdn.co/image/${(_m = spotifyData.preview_url) === null || _m === void 0 ? void 0 : _m.split("?cid=")[1]}`
                            : "https://www.scdn.co/i/_global/twitter_card-default.jpg",
                        duration: Util_1.Util.buildTimeCode(Util_1.Util.parseMS(spotifyData.duration_ms)),
                        views: 0,
                        requestedBy: options.requestedBy,
                        source: "spotify"
                    });
                    return { playlist: null, tracks: [spotifyTrack] };
                }
                case types_1.QueryType.SPOTIFY_PLAYLIST:
                case types_1.QueryType.SPOTIFY_ALBUM: {
                    const spotifyPlaylist = yield spotify_url_info_1.default.getData(query).catch(Util_1.Util.noop);
                    if (!spotifyPlaylist)
                        return { playlist: null, tracks: [] };
                    const playlist = new Playlist_1.Playlist(this, {
                        title: (_o = spotifyPlaylist.name) !== null && _o !== void 0 ? _o : spotifyPlaylist.title,
                        description: (_p = spotifyPlaylist.description) !== null && _p !== void 0 ? _p : "",
                        thumbnail: (_r = (_q = spotifyPlaylist.images[0]) === null || _q === void 0 ? void 0 : _q.url) !== null && _r !== void 0 ? _r : "https://www.scdn.co/i/_global/twitter_card-default.jpg",
                        type: spotifyPlaylist.type,
                        source: "spotify",
                        author: spotifyPlaylist.type !== "playlist"
                            ? {
                                name: (_t = (_s = spotifyPlaylist.artists[0]) === null || _s === void 0 ? void 0 : _s.name) !== null && _t !== void 0 ? _t : "Unknown Artist",
                                url: (_w = (_v = (_u = spotifyPlaylist.artists[0]) === null || _u === void 0 ? void 0 : _u.external_urls) === null || _v === void 0 ? void 0 : _v.spotify) !== null && _w !== void 0 ? _w : null
                            }
                            : {
                                name: (_0 = (_y = (_x = spotifyPlaylist.owner) === null || _x === void 0 ? void 0 : _x.display_name) !== null && _y !== void 0 ? _y : (_z = spotifyPlaylist.owner) === null || _z === void 0 ? void 0 : _z.id) !== null && _0 !== void 0 ? _0 : "Unknown Artist",
                                url: (_3 = (_2 = (_1 = spotifyPlaylist.owner) === null || _1 === void 0 ? void 0 : _1.external_urls) === null || _2 === void 0 ? void 0 : _2.spotify) !== null && _3 !== void 0 ? _3 : null
                            },
                        tracks: [],
                        id: spotifyPlaylist.id,
                        url: (_5 = (_4 = spotifyPlaylist.external_urls) === null || _4 === void 0 ? void 0 : _4.spotify) !== null && _5 !== void 0 ? _5 : query,
                        rawPlaylist: spotifyPlaylist
                    });
                    if (spotifyPlaylist.type !== "playlist") {
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        playlist.tracks = spotifyPlaylist.tracks.items.map((m) => {
                            var _a, _b, _c, _d, _e, _f, _g, _h;
                            const data = new Track_1.default(this, {
                                title: (_a = m.name) !== null && _a !== void 0 ? _a : "",
                                description: (_b = m.description) !== null && _b !== void 0 ? _b : "",
                                author: (_d = (_c = m.artists[0]) === null || _c === void 0 ? void 0 : _c.name) !== null && _d !== void 0 ? _d : "Unknown Artist",
                                url: (_f = (_e = m.external_urls) === null || _e === void 0 ? void 0 : _e.spotify) !== null && _f !== void 0 ? _f : query,
                                thumbnail: (_h = (_g = spotifyPlaylist.images[0]) === null || _g === void 0 ? void 0 : _g.url) !== null && _h !== void 0 ? _h : "https://www.scdn.co/i/_global/twitter_card-default.jpg",
                                duration: Util_1.Util.buildTimeCode(Util_1.Util.parseMS(m.duration_ms)),
                                views: 0,
                                requestedBy: options.requestedBy,
                                playlist,
                                source: "spotify"
                            });
                            return data;
                        });
                    }
                    else {
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        playlist.tracks = spotifyPlaylist.tracks.items.map((m) => {
                            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
                            const data = new Track_1.default(this, {
                                title: (_a = m.track.name) !== null && _a !== void 0 ? _a : "",
                                description: (_b = m.track.description) !== null && _b !== void 0 ? _b : "",
                                author: (_d = (_c = m.track.artists[0]) === null || _c === void 0 ? void 0 : _c.name) !== null && _d !== void 0 ? _d : "Unknown Artist",
                                url: (_f = (_e = m.track.external_urls) === null || _e === void 0 ? void 0 : _e.spotify) !== null && _f !== void 0 ? _f : query,
                                thumbnail: (_j = (_h = (_g = m.track.album) === null || _g === void 0 ? void 0 : _g.images[0]) === null || _h === void 0 ? void 0 : _h.url) !== null && _j !== void 0 ? _j : "https://www.scdn.co/i/_global/twitter_card-default.jpg",
                                duration: Util_1.Util.buildTimeCode(Util_1.Util.parseMS(m.track.duration_ms)),
                                views: 0,
                                requestedBy: options.requestedBy,
                                playlist,
                                source: "spotify"
                            });
                            return data;
                        });
                    }
                    return { playlist: playlist, tracks: playlist.tracks };
                }
                case types_1.QueryType.SOUNDCLOUD_PLAYLIST: {
                    const data = yield soundcloud.getPlaylist(query).catch(Util_1.Util.noop);
                    if (!data)
                        return { playlist: null, tracks: [] };
                    const res = new Playlist_1.Playlist(this, {
                        title: data.title,
                        description: (_6 = data.description) !== null && _6 !== void 0 ? _6 : "",
                        thumbnail: (_7 = data.thumbnail) !== null && _7 !== void 0 ? _7 : "https://soundcloud.com/pwa-icon-192.png",
                        type: "playlist",
                        source: "soundcloud",
                        author: {
                            name: (_11 = (_9 = (_8 = data.author) === null || _8 === void 0 ? void 0 : _8.name) !== null && _9 !== void 0 ? _9 : (_10 = data.author) === null || _10 === void 0 ? void 0 : _10.username) !== null && _11 !== void 0 ? _11 : "Unknown Artist",
                            url: (_12 = data.author) === null || _12 === void 0 ? void 0 : _12.profile
                        },
                        tracks: [],
                        id: `${data.id}`,
                        url: data.url,
                        rawPlaylist: data
                    });
                    for (const song of data.tracks) {
                        const track = new Track_1.default(this, {
                            title: song.title,
                            description: (_13 = song.description) !== null && _13 !== void 0 ? _13 : "",
                            author: (_17 = (_15 = (_14 = song.author) === null || _14 === void 0 ? void 0 : _14.username) !== null && _15 !== void 0 ? _15 : (_16 = song.author) === null || _16 === void 0 ? void 0 : _16.name) !== null && _17 !== void 0 ? _17 : "Unknown Artist",
                            url: song.url,
                            thumbnail: song.thumbnail,
                            duration: Util_1.Util.buildTimeCode(Util_1.Util.parseMS(song.duration)),
                            views: (_18 = song.playCount) !== null && _18 !== void 0 ? _18 : 0,
                            requestedBy: options.requestedBy,
                            playlist: res,
                            source: "soundcloud",
                            engine: song
                        });
                        res.tracks.push(track);
                    }
                    return { playlist: res, tracks: res.tracks };
                }
                case types_1.QueryType.YOUTUBE_PLAYLIST: {
                    const ytpl = yield youtube_sr_1.default.getPlaylist(query).catch(Util_1.Util.noop);
                    if (!ytpl)
                        return { playlist: null, tracks: [] };
                    yield ytpl.fetch().catch(Util_1.Util.noop);
                    const playlist = new Playlist_1.Playlist(this, {
                        title: ytpl.title,
                        thumbnail: ytpl.thumbnail,
                        description: "",
                        type: "playlist",
                        source: "youtube",
                        author: {
                            name: ytpl.channel.name,
                            url: ytpl.channel.url
                        },
                        tracks: [],
                        id: ytpl.id,
                        url: ytpl.url,
                        rawPlaylist: ytpl
                    });
                    playlist.tracks = ytpl.videos.map((video) => {
                        var _a;
                        return new Track_1.default(this, {
                            title: video.title,
                            description: video.description,
                            author: (_a = video.channel) === null || _a === void 0 ? void 0 : _a.name,
                            url: video.url,
                            requestedBy: options.requestedBy,
                            thumbnail: video.thumbnail.url,
                            views: video.views,
                            duration: video.durationFormatted,
                            raw: video,
                            playlist: playlist,
                            source: "youtube"
                        });
                    });
                    return { playlist: playlist, tracks: playlist.tracks };
                }
                default:
                    return { playlist: null, tracks: [] };
            }
        });
    }
    /**
     * Registers extractor
     * @param {string} extractorName The extractor name
     * @param {ExtractorModel|any} extractor The extractor object
     * @param {boolean} [force=false] Overwrite existing extractor with this name (if available)
     * @returns {ExtractorModel}
     */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    use(extractorName, extractor, force = false) {
        if (!extractorName)
            throw new PlayerError_1.PlayerError("Cannot use unknown extractor!", PlayerError_1.ErrorStatusCode.UNKNOWN_EXTRACTOR);
        if (this.extractors.has(extractorName) && !force)
            return this.extractors.get(extractorName);
        if (extractor instanceof ExtractorModel_1.ExtractorModel) {
            this.extractors.set(extractorName, extractor);
            return extractor;
        }
        for (const method of ["validate", "getInfo"]) {
            if (typeof extractor[method] !== "function")
                throw new PlayerError_1.PlayerError("Invalid extractor data!", PlayerError_1.ErrorStatusCode.INVALID_EXTRACTOR);
        }
        const model = new ExtractorModel_1.ExtractorModel(extractorName, extractor);
        this.extractors.set(model.name, model);
        return model;
    }
    /**
     * Removes registered extractor
     * @param {string} extractorName The extractor name
     * @returns {ExtractorModel}
     */
    unuse(extractorName) {
        if (!this.extractors.has(extractorName))
            throw new PlayerError_1.PlayerError(`Cannot find extractor "${extractorName}"`, PlayerError_1.ErrorStatusCode.UNKNOWN_EXTRACTOR);
        const prev = this.extractors.get(extractorName);
        this.extractors.delete(extractorName);
        return prev;
    }
    /**
     * Generates a report of the dependencies used by the `@discordjs/voice` module. Useful for debugging.
     * @returns {string}
     */
    scanDeps() {
        return voice_1.generateDependencyReport();
    }
    /**
     * Resolves queue
     * @param {GuildResolvable|Queue} queueLike Queue like object
     * @returns {Queue}
     */
    resolveQueue(queueLike) {
        return this.getQueue(queueLike instanceof Queue_1.Queue ? queueLike.guild : queueLike);
    }
    *[Symbol.iterator]() {
        yield* Array.from(this.queues.values());
    }
}
exports.Player = Player;
